File: modules/stego-core/src/main/java/com/shterneregen/securelan/stego/service/impl/internal/StegoHeader.java
package com.shterneregen.securelan.stego.service.impl.internal;

import com.shterneregen.securelan.stego.model.StegoContentType;

import java.nio.ByteBuffer;
import java.util.Objects;

public record StegoHeader(StegoContentType contentType, boolean encrypted, int payloadLength) {
    public static final int BYTE_LENGTH = 12;

    private static final int MAGIC = 0x534C5347;
    private static final byte VERSION = 1;
    private static final byte ENCRYPTED_FLAG = 0x01;

    public StegoHeader {
        Objects.requireNonNull(contentType, "contentType");
        if (payloadLength < 0) {
            throw new IllegalArgumentException("payloadLength must not be negative");
        }
    }

    public byte[] write() {
        return ByteBuffer.allocate(BYTE_LENGTH)
                .putInt(MAGIC)
                .put(VERSION)
                .put(encrypted ? ENCRYPTED_FLAG : (byte) 0)
                .put(contentType.code())
                .put((byte) 0)
                .putInt(payloadLength)
                .array();
    }

    public static StegoHeader read(byte[] bytes) {
        Objects.requireNonNull(bytes, "bytes");
        if (bytes.length != BYTE_LENGTH) {
            throw new IllegalArgumentException("Invalid steganography header length");
        }
        ByteBuffer buffer = ByteBuffer.wrap(bytes);
        int magic = buffer.getInt();
        if (magic != MAGIC) {
            throw new IllegalArgumentException("No SecureLanSuite steganography payload found");
        }
        byte version = buffer.get();
        if (version != VERSION) {
            throw new IllegalArgumentException("Unsupported steganography payload version: " + version);
        }
        byte flags = buffer.get();
        boolean encrypted = (flags & ENCRYPTED_FLAG) == ENCRYPTED_FLAG;
        byte contentType = buffer.get();
        buffer.get();
        int payloadLength = buffer.getInt();
        return new StegoHeader(StegoContentType.fromCode(contentType), encrypted, payloadLength);
    }
}

File: modules/stego-core/src/main/kotlin/com/shterneregen/securelan/stego/service/impl/internal/StegoHeader.kt
package com.shterneregen.securelan.stego.service.impl.internal

import com.shterneregen.securelan.stego.model.StegoContentType
import java.nio.ByteBuffer
import java.util.Objects

@JvmRecord
data class StegoHeader(
    val contentType: StegoContentType,
    val encrypted: Boolean,
    val payloadLength: Int,
) {
    init {
        Objects.requireNonNull(contentType, "contentType")
        require(payloadLength >= 0) { "payloadLength must not be negative" }
    }

    fun write(): ByteArray = ByteBuffer.allocate(BYTE_LENGTH)
        .putInt(MAGIC)
        .put(VERSION)
        .put(if (encrypted) ENCRYPTED_FLAG else 0.toByte())
        .put(contentType.code())
        .put(0.toByte())
        .putInt(payloadLength)
        .array()

    companion object {
        const val BYTE_LENGTH: Int = 12

        private const val MAGIC = 0x534C5347
        private const val VERSION: Byte = 1
        private const val ENCRYPTED_FLAG: Byte = 0x01

        @JvmStatic
        fun read(bytes: ByteArray): StegoHeader {
            Objects.requireNonNull(bytes, "bytes")
            if (bytes.size != BYTE_LENGTH) {
                throw IllegalArgumentException("Invalid steganography header length")
            }
            val buffer = ByteBuffer.wrap(bytes)
            val magic = buffer.int
            if (magic != MAGIC) {
                throw IllegalArgumentException("No SecureLanSuite steganography payload found")
            }
            val version = buffer.get()
            if (version != VERSION) {
                throw IllegalArgumentException("Unsupported steganography payload version: $version")
            }
            val flags = buffer.get()
            val encrypted = (flags.toInt() and ENCRYPTED_FLAG.toInt()) == ENCRYPTED_FLAG.toInt()
            val contentType = buffer.get()
            buffer.get()
            val payloadLength = buffer.int
            return StegoHeader(StegoContentType.fromCode(contentType), encrypted, payloadLength)
        }
    }
}
